src/subgraph/service.ts

```diff
@@ -0,0 +1,822 @@
+
+  /**
+   * Converts CommP to hex string
+   */
+  private commPToHex (commP: CommP): string {
+    return toHex(CID.parse(commP.toString()).bytes)
+  }
+
+  /**
+   * Transforms provider data to ApprovedProviderInfo
+   */
+  private transformProviderData (data: any): ApprovedProviderInfo {
+    return {
+      owner: data.address != null && data.address !== '' ? data.address : data.id,
+      pdpUrl: data.pdpUrl,
+      pieceRetrievalUrl: data.pieceRetrievalUrl,
+      registeredAt: this.parseTimestamp(data.registeredAt),
+      approvedAt: this.parseTimestamp(data.approvedAt)
+    }
+  }
+
+  /**
+   * Safely parses timestamp values
+   */
+  private parseTimestamp (value?: number | string): number {
+    if (value == null) return 0
+    const parsed = Number(value)
+    return isNaN(parsed) || parsed === 0 ? 0 : parsed
+  }
+
+  /**
+   * Validates provider data completeness
+   */
+  private isValidProviderData (data: any): boolean {
+    return (
+      data?.id != null &&
+      data.id.trim() !== '' &&
+      data?.pdpUrl != null &&
+      data.pdpUrl.trim() !== '' &&
+      data?.pieceRetrievalUrl != null &&
+      data.pieceRetrievalUrl.trim() !== ''
+    )
+  }
+
+  /**
+   * Queries the subgraph to find approved storage providers that have a specific piece (CommP).
+   *
+   * It sends a GraphQL query to the configured endpoint and parses the response to extract
+   * a list of providers, including their addresses and retrieval URLs.
+   *
+   * @param commp - The piece commitment (CommP) to search for.
+   * @returns A promise that resolves to an array of `ApprovedProviderInfo` objects.
+   *          Returns an empty array if no providers are found or if an error occurs during the fetch.
+   */
+  async getProvidersForCommP (commP: CommP): Promise<ApprovedProviderInfo[]> {
+    const hexCommP = this.commPToHex(commP)
+
+    const data = await this.executeQuery<{ roots: any[] }>(
+      QUERIES.GET_APPROVED_PROVIDERS_FOR_COMMP,
+      { cid: hexCommP },
+      'getProvidersForCommP'
+    )
+
+    if (data?.roots == null || data.roots.length === 0) {
+      console.log(`SubgraphService: No providers found for CommP: ${commP.toString()}`)
+      return []
+    }
+
+    return data.roots
+      .map((root) => root.proofSet.owner)
+      .filter((provider) => {
+        const isValid = this.isValidProviderData(provider)
+        if (!isValid) {
+          console.warn('SubgraphService: Skipping incomplete provider data:', provider)
+        }
+        return isValid
+      })
+      .map((provider) => this.transformProviderData(provider))
+  }
+
+  /**
+   * Queries the subgraph to find a specific approved storage provider by their address.
+   *
+   * @param address - The wallet address of the provider to search for.
+   * @returns A promise that resolves to an `ApprovedProviderInfo` object if the provider is found, or `null` otherwise.
+   */
+  async getProviderByAddress (address: string): Promise<ApprovedProviderInfo | null> {
```

Comment:
is there any requirement on the address format here? does it need to be `0x` prefixed? does it matter? should you add some simple validation and/or docs about that?